distage/distage-testkit/src/main/scala/com/github/pshirshov/izumi/distage/testkit/TestComponentsLifecycleManager.scala

```diff
@@ -0,0 +1,78 @@
+package com.github.pshirshov.izumi.distage.testkit
+
+import java.util.concurrent.ConcurrentLinkedDeque
+
+import com.github.pshirshov.izumi.distage.roles.launcher.ComponentLifecycle
+import com.github.pshirshov.izumi.distage.roles.launcher.exceptions.LifecycleException
+import com.github.pshirshov.izumi.distage.roles.roles.{ComponentsLifecycleManager, RoleComponent}
+import com.github.pshirshov.izumi.distage.testkit.TestComponentsLifecycleManager._
+import com.github.pshirshov.izumi.fundamentals.platform.language.Quirks._
+import com.github.pshirshov.izumi.fundamentals.platform.strings.IzString._
+import com.github.pshirshov.izumi.logstage.api.IzLogger
+
+import scala.collection.JavaConverters._
+import scala.util.{Failure, Try}
+
+class TestComponentsLifecycleManager(
+                                      components: Set[RoleComponent],
+                                      logger: IzLogger,
+                                      resourceCollection: DistageResourceCollection
+                                    ) extends ComponentsLifecycleManager {
+  private val started = new ConcurrentLinkedDeque[ComponentLifecycle]()
+
+  override def componentsNumber: Int = components.size
+
+  private def startComponentWithDeque(component: RoleComponent, deque: ConcurrentLinkedDeque[ComponentLifecycle]): Unit = {
+    deque.push(ComponentLifecycle.Starting(component))
+    component.start()
+    deque.pop().discard()
+    deque.push(ComponentLifecycle.Started(component))
+  }
+
+  override def startComponents(): Unit = {
+    components.foreach { component =>
+      component.synchronized {
+        if (resourceCollection.isMemoized(component)) {
+          if (!isMemoizedComponentStarted(component)) {
```

Comment:
I'd rather put start info into `resourceCollection` too, so that all dirty stuff is in one place

Reply:
@kaishh we need to rewrite this `ComponentLifecycleManager` due to synchronization over components, other `ComponentLifecycle` collection.

Reply:
It's okay to synchronize and use ConcurrentLinkedDeque in main code. I think it would be good to just replace old impl with this and move `isMemoizedComponentStarted` logic into `ResourceCollection`

Reply:
I think that sharing resource collections with main role starter and binding it with DI is not a good idea. So I've just replaced an old collection with a new one and added a synchronization to be able to override `startComponents()` using memoizing components.

Reply:
Don't share, just add methods to `ResourceCollection` like this:

```scala
trait ResourceCollection {
  def register(started: ComponentLifecycle.Started): Unit =
    if (isMemoized(started.component) && !deque.contains(...))
      deque.put

  def isAlreadyStarted(component: Any): Unit = 
      deque.contains(...)
}
```

And keep an independent deque inside ResourceCollection

Reply:
Hmm, okay I guess, maybe it's not a good idea to use ResourceCollection in main starter (it's supposed to be NullResourceCollection there though)